### PR TITLE
complex-numbers: add __eq__ skeleton

### PR DESCRIPTION
**Pull request by pranasziaukas**
All of the needed method skeletons (`__add__`, `__mul__`, etc.) are provided in the initial solution file except for `__eq__`.

Proposing to add it as well for the sake of consistency.

### CODE REVIEW

**Comment by pranasziaukas** [comment]
> Looks good to me. It’s kind of strange we don’t have a direct test for equality, but instead indirect tests on component equality, but having this in the slug is fine.

I guess a direct test on equality by itself would look like
```
self.assertEqual(ComplexNumber(1, 2), ComplexNumber(1, 2))
```
which is a bit trivial.

Otherwise, equality is tested for all the algebraic operations. I think it's fair enough.

**Comment by yawpitch** [comment]
It would be trivial, but from my quick glance at the tests earlier it appears trivial to also pass all the existing tests without ever defining __eq__. We’re assuming it’s presence through the other tests, but never assuring it’s presence.

**Comment by pranasziaukas** [comment]
> It would be trivial, but from my quick glance at the tests earlier it appears trivial to also pass all the existing tests without ever defining __eq__. We’re assuming it’s presence through the other tests, but never assuring it’s presence.

How can you pass the current tests trivially without defining `__eq__`?

**Comment by pranasziaukas** [comment]
I observed that you can have
```
def __eq__(self, other):
    return True
```
and pass the current tests with other methods being rubbish.

In that case a direct test for complex number (in)equality is useful. What do you think?

**Comment by yawpitch** [comment]
> > It would be trivial, but from my quick glance at the tests earlier it appears trivial to also pass all the existing tests without ever defining **eq**. We’re assuming it’s presence through the other tests, but never assuring it’s presence.
> 
> How can you pass the current tests trivially without defining `__eq__`?

That was my faulty quick read on mobile, sorry; you do indeed have to define `__eq__` (in Python 3, you could get away with just `__cmp__` in Python 2), we just never actually verify that it's a meaningful and correct comparison.

**Comment by pranasziaukas** [comment]
Separate tests are fine by me.

**Comment by cmccandless** [comment]
Sorry, busy week getting ready for a move, and I didn't have the time to chime in on this before it was merged.

I have no problem with what was merged, but I wonder if the equality tests should also include the following:
1. test_inequality_different_real_part
2. test_inequality_different_imaginary_part

@yawpitch @pranasziaukas thoughts?

**Comment by pranasziaukas** [comment]
I believe that the current setup strongly suggests to store parts of a complex number in `self.real` and `self.imaginary` (or something comparable) which results in a simple numerical comparison. In that case, if e.g. `number.real == 1` is true then it's not going to hold for any different value.

So I don't see how you could cheat the current test suite without creating some monstrosity with a custom equality.

TL;DR Looks good to me as it is.

**Comment by yawpitch** [comment]
I'd tend to say the single inequality test is fine, since both values are being changed in the test. Hence if they've botched `__eq__` for either the real or imaginary part the test will fail, which should be sufficient.

**Comment by cmccandless** [comment]
I was thinking more if there was an implementation like this:
```Python
class ComplexNumber(object):
    def __eq__(self, other):
        return self.real == other.real
```

`ComplexNumber(2, 1) == ComplexNumber(2, 3)` should be `False`, but would evaluate to `True`. I might be overthinking it.

**Comment by pranasziaukas** [comment]
Okay, @cmccandless I misunderstood your idea yesterday.

In the case you gave it seems like a better idea to test
```
    def test_inequality_of_complex_numbers(self):
        number = ComplexNumber(1, 2)
        number_different_real_part = ComplexNumber(2, 2)
        number_different_imaginary_part = ComplexNumber(1, 1)
        self.assertNotEqual(number, number_different_real_part)
        self.assertNotEqual(number, number_different_imaginary_part)
```
And maybe that's actually better. 
But honestly I don't know enough about the culture at Exercism to know how deep you wish to go down the rabbit hole :)

**Comment by yawpitch** [comment]
I’d personally say that’s exactly as deep down the rabbit hole as we should go for now.

**Comment by yawpitch** [comment]
@pranasziaukas do you want to add the improved inequality tests that handle both real and imaginary parts? I'd be happy to merge this either way.

**Comment by cmccandless** [comment]
That will require a new PR, as this one has already been merged.

**Comment by yawpitch** [comment]
Woops, should have been more clear, edited my previous.
